src/lib/type_check.ml

```diff
@@ -3594,48 +3595,60 @@ and infer_exp env (E_aux (exp_aux, (l, uannot)) as exp) =
           else annot_exp (E_for (v, inferred_t, inferred_f, checked_step, ord, checked_body)) unit_typ
       | _, _ -> typ_error l "Ranges in foreach overlap"
     end
-  | E_if (cond, then_branch, else_branch) ->
+  | E_if (cond, then_branch, else_branch) -> begin
+      (* Try to infer the type of the condition (this helps in some cases).
```

Comment:
Ah you mean it might infer the type as `true` or `false` rather than just `bool`?

Reply:
By the way, where does it actually check the inferred condition type is `true` or `false`? It seems like Sail currently accepts this:

```
if "foo" then...
```

or

```
if 0x12 then...
```

Is that intentional? It actually compiles and runs with the C output at least since it just compiles to `if ("foo") {` and `if (UINT64_C(0x12)) {`. Seems like that might be an oversight though...

Reply:
Ah, that's quite an oversight. There should be a `subtyp inferred_typ bool_typ` check.